test/tla/ChangRobertsTyped_Test.tla

```diff
@@ -0,0 +1,125 @@
+MAX_N == 6
+
+\* we override Node, as N is not known in advance
+OVERRIDE_Node == { i \in 1..MAX_N: i <= N }
+
+\* initialize constants
+ConstInit ==
+    /\ N \in 2..MAX_N
+    /\ Id \in [ 1..MAX_N -> Int ]
+
+\* The below constraints are copied from ASSUME.
+\* They are not enforced automatically, see issue #69.
+Assumptions ==    
+    /\ Node = DOMAIN Id
+    /\ \A n \in Node: Id[n] >= 0
+    /\ \A m,n \in Node : m # n => Id[m] # Id[n]  \* IDs are unique
+
+InitAndAssumptions ==
+    Init /\ Assumptions
+
+(********************** STATELESS OPERATOR TESTS ******************************)
+
+\* Note that succ(n) is not referring to state variables,
+\* so we can test it in isolation.
+\*
+\* @require("ConstInit")
+\* @testStateless
+Test_succ ==
+    \* This is like a property-based test.
+    \* Note that it is exhaustive (for the range of N).
+    \A n \in Node:
+        succ(n) \in Node
+
+\* We could also use a generator, though it is not very useful in case of integers.
+\*
+\* @testStateless
+Test_succ_gen ==
+    \* A property-based test with generators.
+    \* That looks like the canonical form of PBT.
+    \A n \in { Gen(1) }:
+       (n \in Node) => succ(n) \in Node
+
+(*************************** ACTION TESTS *************************************)
+
+\* Preparing the inputs for the test. Note that this is a step of its own.
+\* This is similar to an initialization predicate.
+Prepare_n0 ==
+    \* the following constraint should be added automatically in the future
+    /\ Assumptions
+    \* let the solver pick some data structures within the bounds
+    \* up to 15 messages
+    /\ msgs = Gen(3 * MAX_N)
+    /\ pc = Gen(MAX_N)
+    /\ initiator = Gen(MAX_N)
+    /\ state = Gen(MAX_N)
+    \* restrict the contents with TypeOK,
+    \* so we don't generate useless data structures
+    /\ TypeOK
+
+\* Assertion that we expect to hold true after firing Action_n0.
+Assert_n0 ==
+    \E n, m \in Node:
+        msgs'[n] = msgs[n] \union {m}
+
+\* Execute the action under test.
+\* Note that we decouple Assert_n0 from TestAction_n0.
+\* The reason is that we always assume that TestAction_n0 always holds,
+\* whereas we may want to see Assert_n0 violated.
+\*
+\* @require("ConstInit")
+\* @require("Prepare_n0")
+\* @ensure("Assert_n0")
+\* @testAction
+TestAction_n0 ==
+    \E self \in Node:
+        n0(self)
+
+\* We expect no winner.
+Assert_noWinner ==
+    \A n \in Node:
+        state[n] /= "won"
+
+(*************************** EXECUTION TESTS **********************************)
+\* Execute a sequence of 5 actions, similar to TestAction_n0.
+\* We test a final state with Assert_n0.
+\* Additionally, every state in an execution is tested for Correctness.
+\*
+\* @require("ConstInit")
+\* @require("Prepare_n0")
+\* @invariant("Correctness")
+\* @ensure("Assert_noWinner")
+\* @testExecution(5)
+TestExec_n0_n1 ==
+    \* in this test, we only execute actions by processes 1 and 2
+    \E self \in { 1, 2 }:
+        n0(self) \/ n1(self)
```

Comment:
I really like this idea!

I wonder if it would be possible to somehow write the following:
```tla
IfExec(trace) ==
   /\ trace[0].action.name = "create_client"
   /\ trace[0].action.id = 1
   /\ trace[0].action.height = 3
   /\ trace[1].action.name = "update_client"
   /\ trace[1].action.id = 1
   /\ trace[1].action.height = 2

ThenExec(trace) ==
   /\ trace[0].outcome = "ok" 
   /\ trace[1].outcome = "invalid_height"

\* @require("IfExec")
\* @ensure("ThenExec")
\* @testExecution(2, "Init", "Next")
\* the execution doesn't have to start at `Init`;
\* instead, the execution can start from any state reachable from `Init` by applying `Next`
```

I don't think I'm using `@require` and `@ensure` correctly (i.e. as in the other examples), but I believe that it conveys the idea.

If something like this was supported, it could probably also be used to generate MBT tests in a much nicer way.

Reply:
Do you like to refer to the intermediate states in a trace? The idea was the you can talk about the variables of the first state and the last state. Otherwise, we are quickly approaching temporal logic :)

Reply:
@andrey-kuprianov what is your take on that? How far is it from the tests you need in MBT?

Reply:
> Do you like to refer to the intermediate states in a trace?

Yes. My example is bad in that sense. I should have used a trace with more than 2 states.

Reply:
I think @shonfeder and you are asking about more or less the same thing. I don't like the idea of having `trace` explicitly. I guess what you are asking for is a temporal property. Instead of giving an explicit index, we could write `A /\ []B /\ <>C`. This would be the syntax for advanced users :-)

Reply:
Related: `Trace` operator in CommunityModules https://github.com/tlaplus/CommunityModules/blob/4337afb74653cedcc2a24107759bf1613812db34/modules/TLCExt.tla#L42-L56

Reply:
Yep. I don't like the idea of exposing the trace explicitly. Once you do that, the users will write arbitrary code involving traces, which makes it much harder to analyze for symbolic techniques.

Reply:
I am also not sure what to do about the primed and unprimed variables in this test. It would be nice to make it less of a hack.

Reply:
> Yep. I don't like the idea of exposing the trace explicitly. Once you do that, the users will write arbitrary code involving traces, which makes it much harder to analyze for symbolic techniques.

Power to the ~People~ tools. ✊